File: backend/src/infrastructure/security/SanitizationAdapter.js
'use strict';

const path = require('path');

class SanitizationAdapter {
  /**
   * Prevents XSS by escaping HTML special characters.
   */
  static escapeHtml(str) {
    if (typeof str !== 'string') return '';
    return str
      .replace(/&/g, '&amp;')
      .replace(/</g, '&lt;')
      .replace(/>/g, '&gt;')
      .replace(/"/g, '&quot;')
      .replace(/'/g, '&#039;');
  }

  /**
   * Defends against directory path traversal attacks.
   */
  static sanitizeFilePath(userInput, baseDir = process.cwd()) {
    if (typeof userInput !== 'string') {
      throw new Error('Invalid file path argument');
    }
    const resolvedBase = path.resolve(baseDir);
    const resolved = path.resolve(resolvedBase, userInput);
    if (!resolved.startsWith(resolvedBase + path.sep) && resolved !== resolvedBase) {
      throw new Error('Path traversal attempt detected');
    }
    return resolved;
  }

  /**
   * Escapes regular expression special characters to avoid ReDoS / injection.
   */
  static escapeRegex(str) {
    if (typeof str !== 'string') return '';
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  }
}

module.exports = SanitizationAdapter;

File: backend/app/infrastructure/security/sanitization_adapter.py
import html
import os
import re
from pathlib import Path

class SanitizationAdapter:
    @staticmethod
    def escape_html(text: str) -> str:
        """
        Escapes HTML special characters for XSS defense.
        """
        if not text or not isinstance(text, str):
            return ""
        return html.escape(text, quote=True)

    @staticmethod
    def sanitize_file_path(user_input: str, base_dir: str = None) -> str:
        """
        Sanitizes file path to prevent directory traversal attacks.
        """
        if not user_input or not isinstance(user_input, str):
            raise ValueError("Invalid file path argument")
        base = Path(base_dir or os.getcwd()).resolve()
        target = (base / user_input).resolve()
        if not str(target).startswith(str(base)):
            raise ValueError("Path traversal attempt detected")
        return str(target)

    @staticmethod
    def escape_regex(text: str) -> str:
        """
        Escapes special regex characters to prevent ReDoS injection.
        """
        if not text or not isinstance(text, str):
            return ""
        return re.escape(text)
